### DISCUSSION TITLE
Fallback to a default value instead of ValidationError

### DISCUSSION THREAD

**Discussion by elrik75** [pydantic/pydantic#2741] [Question] [answered]
I have some metrics created by sensors, these sensors are not precise and sometime create bad values.
I would like to exclude bad metrics with Pydantic but accept all other ones by putting default values into the failing metrics.
Is there a elegant solution?

Creating a new config parameter to do that automatically may be useful maybe:
```py
class Metric(BaseModel):
    metricA: confloat(ge=0, le=1) = 0
    metricB: confloat(ge=0, le=1) = 0
    [...]
    metricZ: Optional[confloat(ge=0, le=1)] = None

    class Config:
        # proposition:
        failure_fallbacks_to_default = True
```

Maybe we someone would prefer to set if there is a fallback field by field, then we could use the `Field` function:
```py
class Metric(BaseModel):
    metricA: confloat(ge=0, le=1) = Field(0, failure_fallbacks_to_default=True)

    # will trigger a ValidationError on a bad metric value
    metricB: confloat(ge=0, le=1) = 0

    [...]
    metricZ: Optional[confloat(ge=0, le=1)] = Field(None, failure_fallbacks_to_default=True)
```

What do you think? Did I miss something already possible to solve that?

**Comment by PrettyWood** [comment]
Hi @elrik75 
It's an interesting idea and could be easily added to _pydantic_. Don't know if many people would be interested by it
In the meantime you could go with something like this
```py
from typing import Optional

from pydantic import BaseModel, confloat, validate_model

class Metric(BaseModel):
    metricA: confloat(ge=0, le=1) = 0
    metricB: confloat(ge=0, le=1) = 0
    metricZ: Optional[confloat(ge=0, le=1)] = None


    @classmethod
    def parse_obj_with_default_fallback(cls: type['Metric'], obj: dict[str, object]) -> 'Metric':
        """
        First try to validate all values
        If any value is not valid, it won't be part of the data used in `__init__`.
        If a value is hence missing and required, it will still fail
        """
        valid_values, fields_set, validation_error = validate_model(cls, obj)
        return Metric(**valid_values)

assert Metric.parse_obj_with_default_fallback(dict(metricA=-1)).metricA == 0
assert Metric.parse_obj_with_default_fallback(dict(metricA=0.5)).metricA == 0.5
```
Hope it helps!

**Reply by elrik75** [reply]
I use Pydantic from FastAPI and I don't really want to override the validation system. So it would be helpful if it works in pure Pydantic (I mean without calling an extra method). Do you think I should create an issue?

**Comment by 0xDAFE** [comment]
Hi @PrettyWood, @elrik75 ,
I'm in a similar situation but need the same fallback behaviour for a more complex model with **nested** fields: If any field somewhere deep down the model throws a ValidationError, this approach breaks. The problem is that `valid_values` excludes the nested model if there is a validation error somewhere inside of it, even if other parts would be "good".

```python
from typing import Optional

from pydantic import BaseModel, confloat, validate_model


class NestedMetric(BaseModel):
    nestedN: confloat(ge=0, le=1) = 0
    nestedM: confloat(ge=0, le=1) = 0


class Metric(BaseModel):
    metricA: confloat(ge=0, le=1) = 0
    metricB: confloat(ge=0, le=1) = 0
    metricZ: Optional[confloat(ge=0, le=1)] = None
    nested: NestedMetric

    @classmethod
    def parse_obj_with_default_fallback(cls: type['Metric'], obj: dict[str, object]) -> 'Metric':
        """
        First try to validate all values
        If any value is not valid, it won't be part of the data used in `__init__`.
        If a value is hence missing and required, it will still fail
        """
        valid_values, fields_set, validation_error = validate_model(cls, obj)
        # Note: valid_values does not include the part of "nested" that would be valid
        # nested is either completely valid or not
        print(valid_values)
        fallback = Metric(**valid_values)
        return fallback


valid_nested = {
    "metricA": 0,
    "nested": {
        "nestedN": 0.1
    }
}

invalid_nested = {
    "metricA": 0,
    "nested": {
        "nestedN": 0.1,
        "nestedM": -1
    }
}

assert Metric.parse_obj_with_default_fallback(valid_nested).metricA == 0
assert Metric.parse_obj_with_default_fallback(invalid_nested).nested.nestedM == 0  # doesn't work
```

Any ideas how to solve this in an elegant way?

**Reply by PrettyWood** [reply]
Hi @0xDAFE 
Can't you just add a default value to your field `nested: NestedMetric = NestedMetric()`?

**Reply by 0xDAFE** [reply]
In theory yes. However, the model I am using in practice is a lot more complicated and nests several layers deep. The goal would be not to loose all the other existing and valid information if just one of the fields down there fails to validate.

**Comment by PrettyWood** [comment, marked as answer]
I would go with something like this then
```py
from pydantic import BaseModel as PydanticModel
from pydantic.fields import ModelField


def generate_default_fallback_validate(original_field_validate):
    def default_fallback_validate(self, v, values, *, loc, cls = None):
        value, error = original_field_validate(self, v, values, loc=loc, cls=cls)
        if error and not self.required:
            value, error = self.default, None
        return value, error

    return default_fallback_validate


class BaseModel(PydanticModel):
    @classmethod
    def parse_obj_with_default_fallback(cls, obj):
        original_field_validate = ModelField.validate
        try:
            ModelField.validate = generate_default_fallback_validate(original_field_validate)
            return cls.parse_obj(obj)
        finally:
            ModelField.validate = original_field_validate
```

**Reply by 0xDAFE** [reply]
Thanks for the suggestion! I will give it a shot. Can you think of any pitfalls when patching `ModelField.validate` like this?

**Reply by PrettyWood** [reply]
Honestly this should probably be a feature in pydantic. I think it's fine to patch it like this but there may be some bugs as I didn't try it out 😬

**Reply by elrik75** [reply]
Any chance to have this behavior by setting an option in `Config` ? Because some of us use Pydantic indirectly through another library (FastAPI for ex.) and calling a classmethod will not be easy.

**Comment by Lacrymology** [comment]
I need something similar, but even simpler. I need a string type that if it receives `None` it turns it into a fallback (I have two distinct needs, one for `""` and another for `"INVALID"`, don't ask).

**Reply by 0xDAFE** [reply]
This sounds like it could be solved with a pre-validator (and potentially [re-using](https://pydantic-docs.helpmanual.io/usage/validators/#reuse-validators) it)

Something like this (not tested):
```python
@validator('string_field', pre=True)
def validate_string(cls, v):
    if not v or v == "INVALID":
       return "fallback"
    else:
       return v
```

**Comment by Lacrymology** [comment]
Yes I found this yesterday, thank you 🙂
